### ISSUE TITLE
Hoisting fails

### ISSUE THREAD

**Issue by omphalos** [int3/metajs#12]
The following test throws an exception, but shouldn't:
x(); function x() { console.log(1); }

**Comment by int3** [comment]
Yeah, I was lazy about implementing another AST traversal to extract out the function declarations. We should do that. This also throws when it shouldn't, because we don't hoist var declarations either: `console.log(a); var a;`. I think it'd be nice to use an existing traversal library like `estraverse` rather than writing another huge switch-case statement.

**Comment by michaelficarra** [comment]
@int3: I recommend both [Constellation/estraverse](https://github.com/Constellation/estraverse) and [Constellation/escope](https://github.com/Constellation/escope). My favourite program doesn't work:

``` js
(function(){
  try { throw 0; } catch(a) { var a = 1; }
  console.log(a);
}());
```

**Comment by int3** [comment]
@michaelficarra I actually have [another interpreter](https://github.com/int3/closure-interpreter) that implements hoisting (and several other features) correctly. It does use estraverse, but not escope -- I'll definitely have a look at the latter the next time I write another JS interpreter / tool. closure-interpreter isn't in CPS at the moment, but I think it should be quite easy to hack it into shape, and then hook it up to metajs' frontend. Maybe this weekend.

**Comment by kumavis** [comment]
+1
